### PR TITLE
Fix Report other mouse buttons #4754

### PR DESCRIPTION
**Pull request by marianomarciello**
This enables mapping mouse buttons in mouse_bindings and also enables mouse buttons signal in vim/nvim. So it should be possibile to map buttons X1 and X2 as stated in the vim documentation or at least to have the same behavior of other terminals.

### CODE REVIEW

File: alacritty/src/input.rs [hunk 1]
```diff
@@ -623,8 +622,7 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
                 MouseButton::Left => 0,
                 MouseButton::Middle => 1,
                 MouseButton::Right => 2,
-                // Can't properly report more than three buttons.
-                MouseButton::Other(_) => return,
+                MouseButton::Other(x) => x.try_into().unwrap(),
```
Comment (chrisduerr) [inline alacritty/src/input.rs:625]: I really doubt this reports mouse buttons correctly. Did you just guess at what value should be reported?

**Review by chrisduerr — CHANGES_REQUESTED** [review]

**Comment by alex-courtis** [comment]
Tested with the following:

```yaml
mouse_bindings:
  - { mouse: Left, chars: 'L' }
  - { mouse: Middle, chars: 'M' }
  - { mouse: Right, chars: 'R' }
  - { mouse: 4, chars: '4' }
  - { mouse: 5, chars: '5' }
  - { mouse: 6, chars: '6' }
  - { mouse: 7, chars: '7' }
  - { mouse: 8, chars: '8' }
  - { mouse: 9, chars: '9' }
```

LMR were sent however numbered were not.

**Comment by chrisduerr** [comment]
https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Other-buttons

**Comment by alex-courtis** [comment]
> https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Other-buttons
> 
> For terminal emulators it is important we're reporting the **exact same** things as every other terminal emulator. I recommend testing XTerm for comparison.

FYI I used st to debug the standard xterm control sequences: https://github.com/alacritty/alacritty/issues/4754#issuecomment-939402976

File: alacritty/src/input.rs (hunk 1 above)
Comment (marianomarciello) [reply alacritty/src/input.rs:625]: Hi @chrisduerr  thanks for your time. I have tested this with my mouse that has one additional button (button 8) and the number reported is 8. Then I used xbindkeys to bind button 9, 10 and 11 with my left shift and the value reported was 9, 10 and 11 respectively.  Should I follow the convention linked by you ? (https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Other-buttons)

Comment (chrisduerr) [reply alacritty/src/input.rs:625]: Yes you should.

**Comment by alex-courtis** [comment]
> Can you provide me the mapping of your buttons ? (You can use xev for this).

Wayland and X11 have different behaviour regarding button numbers.

X11:
- xev reports button press 1 to 9 as expected
  - 1 : left
  - 2 : middle
  - 3 : right
  - 4 : wheel up
  - 5 : wheel down
  - 6 : wheel left
  - 7 : wheel right
  - 8 : side rear
  - 9 : side front
- Buttons 8 to 9 are seen by alacritty and send 8, 9
- Buttons 4 to 7 are NOT seen by alacritty

Wayland:
- wev reports differently to xev
  - 272(left) : left
  - 274(middle) : middle
  - 273(right) : right
  - axis events : wheel up/down/left/right
  - 275(side) : side rear
  - 276(extra) : side front
- xev reports same as under X11
- Buttons 275 and 276 are seen by alacritty and send S, E
- Buttons 4 to 9 are NOT seen by alacritty

This behaviour does not appear to be any different to master.

```yaml
mouse_bindings:
  - { mouse: Left, chars: 'L' }
  - { mouse: Middle, chars: 'M' }
  - { mouse: Right, chars: 'R' }
  - { mouse: 4, chars: '4' }
  - { mouse: 5, chars: '5' }
  - { mouse: 6, chars: '6' }
  - { mouse: 7, chars: '7' }
  - { mouse: 8, chars: '8' }
  - { mouse: 9, chars: '9' }
  - { mouse: 272, chars: 'l' }
  - { mouse: 273, chars: 'r' }
  - { mouse: 274, chars: 'm' }
  - { mouse: 275, chars: 'S' }
  - { mouse: 276, chars: 'E' }
```

Mouses tested: logitech g502, steelseries rival 3. Model does not appear to be relevant.

**Comment by kchibisov** [comment]
Axis events are not send via buttons in the first place by our upstream library, and won't be sent, because it's not a button.

The thing is that every button should be normalized to a value it should, which is not the case and that's why it's questionable. I'm pretty sure on macOS/Windows the situation will be different and that event on X11/Wayland it's behaving not the way it should.

I don't think application generally support mouse buttons other than `back` and `forward` (which is 4 and 5 or 6/7, idk). Thus I'd suggest to translate only them and there's a patch upstream to namely expose such buttons in winit https://github.com/rust-windowing/winit/pull/2770

**Comment by marianomarciello** [comment]
> 

Hi @kchibisov  thanks for your time.
I think `back` and `forward` buttons are not that useful in a terminal. I would prefer to report those buttons in a way that we can then bind them to a specific action (using mouse_bindings). `back` and `forward` can then be enabled using mouse_bindings.

**Comment by marianomarciello** [comment]
> 

Hi @alex-courtis. I think the Mousewheel is handled with the `MouseWheel` field in the `WindowEvent` enum (from the winit crates). While MouseButtons (like button 8..9) are handled using the enum `MouseButton` (from the winint creates). So long story short I think that the Mousewheel events are handled before the MouseButton code, so they can not reach the code that I am changing. I think this can be part of another PR. I will try to fix the value reported (using the xterm convention) and then switch to the MouseWheel issue.

**Comment by alex-courtis** [comment]
> I think this can be part of another PR. I will try to fix the value reported (using the xterm convention) and then switch to the MouseWheel issue.

That's fantastic. Please let me know when you need more testing.

**Comment by kchibisov** [comment]
I'm still not sure how it should all work and translate, surely it might work on X11, but outside of it, I highly doubt it.

However the new winit API defined `Back` and `Forward` on all the platforms and those are `8` and `9` in your PR (the same will map to Wayland).

**Comment by marianomarciello** [comment]
Hi @kchibisov thanks again for your time.
I only tested it in X11, @alex-courtis tested it in wayland too. The `8` that I was talking about is because my hardware has this mapping. So my `Back` button is button 8. I think this is not so rare.  So there is no guarantee that some buttons correspond to some functionality. Maybe it is better to let the user decide the mapping in the configuration file ?

**Comment by kchibisov** [comment]
Well, the thing is that reporting for such buttons will be broken most of the time, because Xterm references X11, where it'll work, because Xorg remaps the buttons, but not on Wayland, macOS, Windows, where there's no such thing.

What I've suggested is to map `Back` and `Forward` to `8` and `9` and report it the way it should according to Xterm, as for the rest, I'm not sure we should report them, because we can't translate them to the way xterm spec expected them to be. (On Wayland you have 0x110+, because those are coming from libinput iirc).

**Comment by alex-courtis** [comment]
> What I've suggested is to map `Back` and `Forward` to `8` and `9`

That appears to be a reliable standard; seen across all mouse hardware I've owned.

**Comment by marianomarciello** [comment]
So we need to avoid reporting 8 and 9 ? Maybe I need to take a look into the button mapping after this PR [https://github.com/rust-windowing/winit/pull/2770](https://github.com/rust-windowing/winit/pull/2770), maybe everything works out of the box after this rework :)

**Comment by kchibisov** [comment]
@marianomarciello the said branch merged, we now get button events out of winit which we should be able to map to `8` and `9` for the mouse mode from what I can say.

If the intention is to use them in the config, you can bind `Back` and `Forward` buttons.

**Comment by chrisduerr** [comment]
Closing the PR since it's dead.

**Comment by VishnuSanal** [comment]
> What I've suggested is to map `Back` and `Forward` to `8` and `9`

@chrisduerr this was the only thing that was missing in this PR, right?

**Comment by chrisduerr** [comment]
https://github.com/alacritty/alacritty/pull/6957#issuecomment-1595028204

**Pull request closed without merging**